crypto/evp/evp_lib.c

```diff
@@ -925,12 +926,32 @@ int (*EVP_MD_meth_get_ctrl(const EVP_MD *md))(EVP_MD_CTX *ctx, int cmd,
     return md->md_ctrl;
 }
 
+#ifndef OPENSSL_NO_DEPRECATED_3_0
 const EVP_MD *EVP_MD_CTX_md(const EVP_MD_CTX *ctx)
 {
     if (ctx == NULL)
         return NULL;
     return ctx->reqdigest;
 }
+#endif
+
+const EVP_MD *EVP_MD_CTX_get0_md(const EVP_MD_CTX *ctx)
+{
+    if (ctx == NULL)
+        return NULL;
+    return ctx->reqdigest;
+}
+
+EVP_MD *EVP_MD_CTX_get1_md(EVP_MD_CTX *ctx)
+{
+    EVP_MD *md;
+
+    if (ctx == NULL)
+        return NULL;
+    md = (EVP_MD *)ctx->reqdigest;
+    ctx->reqdigest = NULL;
```

Comment:
This looks odd.  I would rather have seen it increment the refcount if `md->origin == EVP_ORIG_DYNAMIC`, and *not* clear `ctx->reqdigest`.  What's the reason you did it like this?  What's going to happen if the application called this, and then some internal controls want to call `EVP_MD_CTX_get0_md()` (which, you may have noticed, happens in quite a few places)?

Reply:
Suggested change:

``` diff
diff --git a/crypto/evp/evp_lib.c b/crypto/evp/evp_lib.c
index 3467e4f0e1..aece8cefd4 100644
--- a/crypto/evp/evp_lib.c
+++ b/crypto/evp/evp_lib.c
@@ -949,7 +949,8 @@ EVP_MD *EVP_MD_CTX_get1_md(EVP_MD_CTX *ctx)
     if (ctx == NULL)
         return NULL;
     md = (EVP_MD *)ctx->reqdigest;
-    ctx->reqdigest = NULL;
+    if (!EVP_MD_up_ref(md))
+        md = NULL;
     return md;
 }
 
```

Reply:
(I've tested a merge this PR with #14219, and had a crash because of heap-after-free...  with the change I just suggested, the crash disappeared)

Reply:
Wow, your fix is awesome!  Thanks.